README.md
# Deprecated
This project is deprecated and archived. Please don't use it for anything other than learning purposes.

# sudo-prompt

Run a non-graphical terminal command using `sudo`, prompting the user with a graphical OS dialog if necessary. Useful for background Node.js applications or native Electron apps that need `sudo`.

## Cross-Platform
`sudo-prompt` provides a native OS dialog prompt on **macOS**, **Linux** and **Windows**.

![macOS](https://raw.githubusercontent.com/jorangreef/sudo-prompt/master/macos.png)

![Linux](https://raw.githubusercontent.com/jorangreef/sudo-prompt/master/linux.png)

![Windows](https://raw.githubusercontent.com/jorangreef/sudo-prompt/master/windows.png)

## Installation
`sudo-prompt` has no external dependencies and does not require any native bindings.
```
npm install sudo-prompt
```

## Usage
Note: Your command should not start with the `sudo` prefix.
```javascript
var sudo = require('sudo-prompt');
var options = {
  name: 'Electron',
  icns: '/Applications/Electron.app/Contents/Resources/Electron.icns', // (optional)
};
sudo.exec('echo hello', options,
  function(error, stdout, stderr) {
    if (error) throw error;
    console.log('stdout: ' + stdout);
  }
);
```

`sudo-prompt` will use `process.title` as `options.name` if `options.name` is not provided. `options.name` must be alphanumeric only (spaces are supported) and at most 70 characters.

`sudo-prompt` will preserve the current working directory on all platforms. Environment variables can be set explicitly using `options.env`.

**`sudo-prompt.exec()` is different to `child-process.exec()` in that no child process is returned (due to platform and permissions constraints).**

## Behavior
On macOS, `sudo-prompt` should behave just like the `sudo` command in the shell. If your command does not work with the `sudo` command in the shell (perhaps because it uses `>` redirection to a restricted file), then it may not work with `sudo-prompt`. However, it is still possible to use sudo-prompt to get a privileged shell, [see this closed issue for more information](https://github.com/jorangreef/sudo-prompt/issues/1).

On Linux, `sudo-prompt` will use either `pkexec` or `kdesudo` to show the password prompt and run your command. Where possible, `sudo-prompt` will try and get these to mimic `sudo`. Depending on which binary is used, and due to the limitations of some binaries, the name of your program or the command itself may be displayed to your user. `sudo-prompt` will not use `gksudo` since `gksudo` does not support concurrent prompts. Passing `options.icns` is currently not supported by `sudo-prompt` on Linux. Patches are welcome to add support for icons based on `polkit`.

On Windows, `sudo-prompt` will elevate your command using User Account Control (UAC). Passing `options.name` or `options.icns` is currently not supported by `sudo-prompt` on Windows.

## Non-graphical terminal commands only
Just as you should never use `sudo` to launch any graphical applications, you should never use `sudo-prompt` to launch any graphical applications. Doing so could cause files in your home directory to become owned by root. `sudo-prompt` is explicitly designed to launch non-graphical terminal commands. For more information, [read this post](http://www.psychocats.net/ubuntu/graphicalsudo).

## Concurrency
On systems where the user has opted to have `tty-tickets` enabled (most systems), each call to `exec()` will result in a separate password prompt. Where `tty-tickets` are disabled, subsequent calls to `exec()` will still require a password prompt, even where the user's `sudo` timestamp file remains valid, due to edge cases with `sudo` itself, [see this discussion for more information](https://github.com/jorangreef/sudo-prompt/pull/76).

You should never rely on `sudo-prompt` to execute your calls in order. If you need to enforce ordering of calls, then you should explicitly order your calls in your application. Where your commands are short-lived, you should always queue your calls to `exec()` to make sure your user is not overloaded with password prompts.

## Invalidating the timestamp
On macOS and Linux, you can invalidate the user's `sudo` timestamp file to force the prompt to appear by running the following command in your terminal:

```sh
$ sudo -k
```


CHANGELOG.md
## [9.2.0] 2020-04-29

### Fixed

- Update TypeScript types to accommodate recent changes, see
[#117](https://github.com/jorangreef/sudo-prompt/issues/117).

## 10.0.0 - 2021-11-15

* major: Convert to TS [Lorenzo Alberto Maria Ambrosi]
* patch: Start TS refactor [Lorenzo Alberto Maria Ambrosi]
* patch: Add author [Lorenzo Alberto Maria Ambrosi]
* patch: Remove resinci file [Lorenzo Alberto Maria Ambrosi]
* patch: Add repo.yml [Lorenzo Alberto Maria Ambrosi]
* patch: Use balena namespace for npm [Lorenzo Alberto Maria Ambrosi]
* patch: Add .resinci.yml to enable balenaCI [Lorenzo Alberto Maria Ambrosi]
* Copy cmd.exe if the path contains any of thes characters: ` ' " & | < > ^ [Alexis Svinartchouk]
* Workaround issue #97 by copying cmd.exe into the temporary folder and running it from here [Alexis Svinartchouk]

## 9.2.1 - 2021-11-11

* patch: Add author [Lorenzo Alberto Maria Ambrosi]
* patch: Remove resinci file [Lorenzo Alberto Maria Ambrosi]
* patch: Add repo.yml [Lorenzo Alberto Maria Ambrosi]
* patch: Use balena namespace for npm [Lorenzo Alberto Maria Ambrosi]
* patch: Add .resinci.yml to enable balenaCI [Lorenzo Alberto Maria Ambrosi]
* Copy cmd.exe if the path contains any of thes characters: ` ' " & | < > ^ [Alexis Svinartchouk]
* Workaround issue #97 by copying cmd.exe into the temporary folder and running it from here [Alexis Svinartchouk]

## [9.1.0] 2019-11-13

### Added

- Add TypeScript types.

## [9.0.0] 2019-06-03

### Changed

- Make cross-platform `stdout`, `stderr` behavior consistent, see
[#89](https://github.com/jorangreef/sudo-prompt/issues/89).

- Preserve current working directory on all platforms.

- Improve kdesudo dialog appearance.

### Added

- Add `options.env` to set environment variables on all platforms, see
[#91](https://github.com/jorangreef/sudo-prompt/issues/91).

### Fixed

- Always return PERMISSION_DENIED as an Error object.

- Support multiple commands separated by semicolons on Linux, see
[#39](https://github.com/jorangreef/sudo-prompt/issues/39).

- Distinguish between elevation errors and command errors on Linux, see
[#88](https://github.com/jorangreef/sudo-prompt/issues/88).

- Fix Windows to return `PERMISSION_DENIED` Error even when Windows' error
messages are internationalized, see 
[#96](https://github.com/jorangreef/sudo-prompt/issues/96).

## [8.2.5] 2018-12-12

### Fixed

- Whitelist package.json files.

## [8.2.4] 2018-12-12

### Added

- A CHANGELOG.md file, see
[#78](https://github.com/jorangreef/sudo-prompt/issues/78).

## [8.2.3] 2018-09-11

### Fixed

- README: Link to concurrency discussion.

## [8.2.2] 2018-09-11

### Fixed

- README: Details on concurrency.

## [8.2.1] 2018-09-11

### Fixed

- A rare idempotency edge case where a command might have been run more than
once, given a very specific OS environment setup.

## [8.2.0] 2018-03-22

### Added

- Windows: Fix `cd` when `cwd` is on another drive, see
[#70](https://github.com/jorangreef/sudo-prompt/issues/70).

## [8.1.0] 2018-01-10

### Added

- Linux: Increase `maxBuffer` limit to 128 MiB, see
[#66](https://github.com/jorangreef/sudo-prompt/issues/66).

## [8.0.0] 2018-11-02

### Changed

- Windows: Set code page of command batch script to UTF-8.

## [7.1.1] 2017-07-18

### Fixed

- README: Explicitly mention that no child process is returned.

## [7.0.0] 2017-03-15

### Changed

- Add status code to errors on Windows and macOS.

## [6.2.1] 2016-12-16

### Fixed

- README: Syntax highlighting.

## [6.2.0] 2016-08-17

### Fixed

- README: Rename OS X to macOS.

## [6.1.0] 2016-08-02

### Added

- Yield an error if no polkit authentication agent is found, see
[#29](https://github.com/jorangreef/sudo-prompt/issues/29).

## [6.0.2] 2016-07-21

### Fixed

- README: Update explanation of Linux behavior.

## [6.0.1] 2016-07-15

### Fixed

- Update keywords in package.json.

## [6.0.0] 2016-07-15

### Changed

- Add support for Windows.


tests/classic.ts
import * as assert from 'assert';
import * as fs from 'fs';
import { resolve } from 'path';
import * as sudo from '../lib';
import { exec } from '../lib/utils';

function kill() {
  if (process.platform === 'win32') {
    return;
  }
  exec('sudo -k');
}

function icns() {
  if (process.platform !== 'darwin') {
    return undefined;
  }
  const iconsPath = resolve(__dirname, '../assets/electron.icns');
  try {
    fs.statSync(iconsPath);
    return iconsPath;
  } catch (error) {
    throw error;
  }
}

main();

async function main() {
  const options = {
    env: { 'SUDO_PROMPT_TEST_ENV': 'hello world' },
    icns: icns(),
    name: 'Electron'
  };
  console.log(options);
  let command, expected;
  if (process.platform === 'win32') {
    command = 'echo %SUDO_PROMPT_TEST_ENV%';
    expected = 'hello world\r\n';
  } else {
    // We use double quotes to tell echo to preserve internal space:
    command = 'echo "$SUDO_PROMPT_TEST_ENV"';
    expected = 'hello world\n';
  }
  console.log(
    'sudo.exec(' +
      JSON.stringify(command) + ', ' +
      JSON.stringify(options) +
    ')'
  );
  try { 
    const { stderr, stdout } = await sudo.exec(command, options);
    console.log('stdout: ' + JSON.stringify(stdout));
    console.log('stderr: ' + JSON.stringify(stderr));
    assert(stdout === undefined || typeof stdout === 'string');
    assert(stderr === undefined || typeof stderr === 'string');
    if (stdout !== expected) {
      throw new Error('stdout != ' + JSON.stringify(expected));
    }
    if (stderr !== '') {
      throw new Error('stderr != ""');
    }
    console.log('OK');
    kill();
  } catch (error) {
    console.error(error);
    assert(error === undefined || typeof error === 'object');
  }
}


tests/concurrent.ts
import * as sudo from '../lib';
import { exec } from 'child_process';

function kill(end) {
  if (process.platform === 'win32') return end();
  exec('sudo -k', end);
}
kill(
  function() {
    var options = {
      name: 'Sudo Prompt'
    };
    if (process.platform === 'win32') {
      var sleep = 'timeout /t 10\r\necho world';
    } else {
      var sleep = 'sleep 10 && echo world';
    }
    sudo.exec(sleep, options,
      function(error, stdout, stderr) {
        console.log(error, stdout, stderr);
      }
    );
    sudo.exec('echo hello', options,
      function(error, stdout, stderr) {
        console.log(error, stdout, stderr);
      }
    );
  }
);


lib/utils.ts
import * as child_process from 'child_process';
import * as crypto from 'crypto';
import { promisify } from 'util';
const randomBytes = promisify(crypto.randomBytes);
const createHash = promisify(crypto.createHash);
export const exec = promisify(child_process.exec);

export const sleep = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

export const PERMISSION_DENIED = 'User did not grant permission.';
export const NO_POLKIT_AGENT = 'No polkit authentication agent found.';

// See issue 66:
export const MAX_BUFFER = 134217728;

export type Options = {
  name?: string;
  icns?: string;
  env: NodeJS.ProcessEnv;
  encoding?: string;
  cwd?: string;
}

export interface Instance {
  command: string;
  options: Options;
  uuid: string;
  path: string;
  pathElevate: string;
  pathExecute: string;
  pathCommand: string;
  pathStdout: string;
  pathStderr: string;
  pathStatus: string;
}

export async function UUID(instance: Instance) {
  try  {
    const random: Buffer = await randomBytes(256);
    const hash: crypto.Hash = await createHash('SHA256', {}) as crypto.Hash;
    hash.update('sudo-prompt-3');
    hash.update(instance.options.name ?? '');
    hash.update(instance.command);
    hash.update(random);
    const uuid = hash.digest('hex').slice(-32);
    if (!uuid || typeof uuid !== 'string' || uuid.length !== 32) {
      // This is critical to ensure we don't remove the wrong temp directory.
      throw new Error('Expected a valid UUID.');
    }
    return uuid;
  } catch (error) {
    throw error
  }
}

export async function Remove(path: string) {
  try {
    if (process.platform === 'win32' && /"/.test(path)) {
      throw new Error('Argument path cannot contain double-quotes.');
    }
    const command = [];
    command.push('rmdir /s /q "' + path + '"');
    command.push('/bin/rm');
    command.push('-rf');
    command.push('"' + EscapeDoubleQuotes(path.normalize(path)) + '"');
    const commandString: string = command.join(' ');
    const {stderr} = await exec(commandString, { encoding: 'utf-8' });
    if (stderr) {
      throw new Error(stderr);
    }
  } catch (error) {
    throw error;
  }
}

export function EscapeDoubleQuotes(str: string) {
  if (typeof str !== 'string') throw new Error('Expected a string.');
  return str.replace(/"/g, '\\"');
}

lib/win32/index.ts
import { promises as fs } from 'fs';
import * as os from 'os';
import * as path from 'path';
import { UUID, Remove, PERMISSION_DENIED, Instance, Options, exec, sleep } from '../utils';

export async function Windows(instance: Instance) {
  try {
    const temp = os.tmpdir();
    const uuid = await UUID(instance);
    instance.uuid = uuid;
    instance.path = path.join(temp, instance.uuid);
    if (/"/.test(instance.path)) {
      // We expect double quotes to be reserved on Windows.
      // Even so, we test for this and abort if they are present.
      throw new Error('Invalid path: instance.path cannot contain double-quotes.')
    }
    instance.pathElevate = path.join(instance.path, 'elevate.vbs');
    instance.pathExecute = path.join(instance.path, 'execute.bat');
    instance.pathCommand = path.join(instance.path, 'command.bat');
    instance.pathStdout = path.join(instance.path, 'stdout');
    instance.pathStderr = path.join(instance.path, 'stderr');
    instance.pathStatus = path.join(instance.path, 'status');
    await fs.mkdir(instance.path);
    await Remove(instance.path);
    await WindowsWriteExecuteScript(instance);
    await WindowsWriteCommandScript(instance);
    await WindowsCopyCmd(instance);
    await WindowsElevate(instance);
    await WindowsWaitForStatus(instance);
    return WindowsResult(instance);
  } catch (error) {
    throw error;
  }
}

function WindowsNeedsCopyCmd(instance: Instance): boolean {
  const specialChars = ['&', '`', "'", '"', '<', '>', '|', '^'];
  for (const specialChar of specialChars) {
    if (instance.path.includes(specialChar)) {
      return true;
    }
  }
  return false;
}

async function WindowsCopyCmd(instance: Instance) {
  const needsCopy = await WindowsNeedsCopyCmd(instance);
  if (!needsCopy) {
    return;
  }
  // Work around https://github.com/jorangreef/sudo-prompt/issues/97
  // Powershell can't properly escape amperstands in paths.
  // We work around this by copying cmd.exe in our temporary folder and running
  // it from here (see WindowsElevate below).
  // That way, we don't have to pass the path containing the amperstand at all.
  // A symlink would probably work too but you have to be an administrator in
  // order to create symlinks on Windows.
  await fs.copyFile(
    path.join(process.env.SystemRoot || '', 'System32', 'cmd.exe'),
    path.join(instance.path, 'cmd.exe'),
  );
}

async function WindowsElevate(instance: Instance) {
  try {
    // We used to use this for executing elevate.vbs:
    // const command = 'cscript.exe //NoLogo "' + instance.pathElevate + '"';
    const command = [];
    command.push('powershell.exe');
    command.push('Start-Process');
    command.push('-FilePath');
    const options: Options = { encoding: 'utf8', env: process.env };
    if (WindowsNeedsCopyCmd(instance)) {
      // path.join('.', 'cmd.exe') would return 'cmd.exe'
      command.push(['.', 'cmd.exe'].join(path.sep));
      command.push('-ArgumentList');
      command.push('"/C","execute.bat"');
      options.cwd = instance.path;
    } else {
      // Escape characters for cmd using double quotes:
      // Escape characters for PowerShell using single quotes:
      // Escape single quotes for PowerShell using backtick:
      // See: https://ss64.com/ps/syntax-esc.html
      command.push('"\'' + instance.pathExecute.replace(/'/g, "`'") + '\'"');
    }
    command.push('-WindowStyle hidden');
    command.push('-Verb runAs');
    const commandString = command.join(' ');
    const child = await exec(commandString, options);
    // @ts-ignore - child.stdin is not typed
    child.stdin.end(); // Otherwise PowerShell waits indefinitely on Windows 7.
  } catch (error) {
    throw new Error(PERMISSION_DENIED);
  }
}

async function WindowsResult(instance: Instance) {
  try {
    await fs.readFile(instance.pathStatus, 'utf-8');
    const stdout = await fs.readFile(instance.pathStdout, 'utf-8');
    const stderr = await fs.readFile(instance.pathStderr, 'utf-8');
    return { stdout, stderr };
  } catch (error) {
    const code = parseInt(error.code.trim(), 10);
    if (code === 0) {
      return;
    } else {
      error = new Error(
        'Command failed: ' + instance.command + '\r\n' + error.message,
      );
      error.code = code.toString();
    }
    throw error;
  }
}

async function WindowsWaitForStatus(instance: Instance) {
  try {
    // VBScript cannot wait for the elevated process to finish so we have to poll.
    // VBScript cannot return error code if user does not grant permission.
    // PowerShell can be used to elevate and wait on Windows 10.
    // PowerShell can be used to elevate on Windows 7 but it cannot wait.
    // powershell.exe Start-Process cmd.exe -Verb runAs -Wait
    const stats = await fs.stat(instance.pathStatus);
    if ( stats.size < 2) {
      await retryWaitForStatus(instance);
    }
  } catch (error) {
    if (error && error.code === 'ENOENT') {
      // Retry if file does not exist or is not finished writing.
      // We expect a file size of 2. That should cover at least "0\r".
      // We use a 1 second timeout to keep a light footprint for long-lived
      // sudo-prompt processes.
      await retryWaitForStatus(instance);
    } else {
      throw error;
    }
  }
}

async function retryWaitForStatus(instance: Instance) {
  await sleep(1000);
  await fs.stat(instance.pathStatus);
  await WindowsWaitForStatus(instance);
}

function WindowsWriteCommandScript(instance: Instance) {
  const cwd = process.cwd();
  if (/"/.test(cwd)) {
    // We expect double quotes to be reserved on Windows.
    // Even so, we test for this and abort if they are present.
    throw new Error('process.cwd() cannot contain double-quotes.');
  }
  const script = [];
  script.push('@echo off');
  // Set code page to UTF-8:
  script.push('chcp 65001>nul');
  // Preserve current working directory:
  // We pass /d as an option in case the cwd is on another drive (issue 70).
  script.push('cd /d "' + cwd + '"');
  // Export environment variables:
  for (const key in instance.options.env) {
    // "The characters <, >, |, &, ^ are special command shell characters, and
    // they must be preceded by the escape character (^) or enclosed in
    // quotation marks. If you use quotation marks to enclose a string that
    // contains one of the special characters, the quotation marks are set as
    // part of the environment variable value."
    // In other words, Windows assigns everything that follows the equals sign
    // to the value of the variable, whereas Unix systems ignore double quotes.
    const value = instance.options.env[key];
    if (value) {
      script.push('set ' + key + '=' + value.replace(/([<>\\|&^])/g, '^$1'));
    }
  }
  script.push(instance.command);
  const scriptString = script.join('\r\n');
  return fs.writeFile(instance.pathCommand, scriptString, 'utf-8');
}

// @ts-ignore
function WindowsWriteElevateScript(instance: Instance) {
  // We do not use VBScript to elevate since it does not return an error if
  // the user does not grant permission. This is here for reference.
  // const script = [];
  // script.push('Set objShell = CreateObject("Shell.Application")');
  // script.push(
  // 'objShell.ShellExecute "' + instance.pathExecute + '", "", "", "runas", 0'
  // );
  // script = script.join('\r\n');
  // fs.writeFile(instance.pathElevate, script, 'utf-8');
}

function WindowsWriteExecuteScript(instance: Instance) {
  const script = [];
  script.push('@echo off');
  script.push(
    'call "' + instance.pathCommand + '"' +
    ' > "' + instance.pathStdout + '" 2> "' + instance.pathStderr + '"'
  );
  script.push('(echo %ERRORLEVEL%) > "' + instance.pathStatus + '"');
  const scriptString = script.join('\r\n');
  fs.writeFile(instance.pathExecute, scriptString, 'utf-8');
}